### ISSUE TITLE
Finetune in scan118 error

### ISSUE THREAD

**Issue by jerryxu9905** [xxlong0/SparseNeuS#17]
Hi, I have succeeded to extract correct mesh in **Generic mode**, but in **Finetune mode**, I found the extracted mesh become confusing during training.

At the begining of Finetune (**0 iters**):
![scan118](https://user-images.githubusercontent.com/97662087/188343173-045eee08-314f-4989-810a-70b18fa37104.png)

End of Finetune (**11k iters**):
![scan118_2](https://user-images.githubusercontent.com/97662087/188343299-52c2a85d-e9c8-4691-be16-b8400ff39425.png)

It seems that the Finetune in scan118 degenerate, I don't modify any config or source code, so what's problem? Can anyone who encounter same issue share some solution? Many Thanks!

**Comment by flamehaze1115** [comment]
Hello. Does the "weight_sum" become 0 in fine-tuning?
If so, you can adjust the two parameters "visibility_beta, and visibility_gama" used in fine-tuning.

**Comment by youmi-zym** [comment]
Hi @flamehaze1115 @xxlong0 @jerryxu9905 ,
I find another bug that happens at [dtu_fit.py](https://github.com/xxlong0/SparseNeuS/blob/main/data/dtu_fit.py#L223):

```python
if self.split == 'train':
            render_idx = self.train_img_idx[idx % self.n_views]
            support_idxs = [render_idx]
```
The query/reference and src views are the same, which means whatever the [sdf_network.color_blend](https://github.com/xxlong0/SparseNeuS/blob/main/models/sparse_neus_renderer.py#L283) predicts, the network will always output the ground-truth color. As a result, there is no supervision for SDF and depth. 
For example, for points along the same ray in the reference view, as the intrinsic and extrinsic parameters of reference and src views are the same,  the points along the ray will be projected to the same pixel in the reference image, whatever the depth is. Therefore, each ray after pixel warping takes the ground truth color and the photometric loss will be 0.

The solution I tried is:
```python
if self.split == 'train':
            render_idx = self.train_img_idx[idx % self.n_views]
            support_idxs = [i for i in self.train_img_idx if i != render_idx]
```
Even though it didn't solve the problem during fine-tuning, it works in my refactored code (based on your framework and pixel/patch-based loss). 

Hope my finding could help, correct me if I'm wrong! Thanks.

**Comment by xxlong0** [comment]
You are right youmi. This is the bug that cause finetuing fails. Sorry for the inconvenience. I push the codes rushly, i should have checked the codes carefully to avoid the inconvenience.

Best
Xiaoxiao
________________________________
发件人: youmi ***@***.***>
发送时间: Tuesday, September 13, 2022 11:46:13 PM
收件人: xxlong0/SparseNeuS ***@***.***>
抄送: Long Xiaoxiao ***@***.***>; Mention ***@***.***>
主题: Re: [xxlong0/SparseNeuS] Finetune in scan118 error (Issue #17)


Hi @flamehaze1115<https://github.com/flamehaze1115> @xxlong0<https://github.com/xxlong0> @jerryxu9905<https://github.com/jerryxu9905> ,
I find another bug that happens at dtu_fit.py<https://github.com/xxlong0/SparseNeuS/blob/main/data/dtu_fit.py#L223>:

if self.split == 'train':
            render_idx = self.train_img_idx[idx % self.n_views]
            support_idxs = [render_idx]

The query/reference and src views are the same, which means whatever the sdf_network.color_blend<https://github.com/xxlong0/SparseNeuS/blob/main/models/sparse_neus_renderer.py#L283> predicts, the network will always output the ground-truth color. As a result, there is no supervision for SDF and depth.
For example, for points along the same ray in the reference view, as the intrinsic and extrinsic parameters of reference and src views are the same, the points along the ray will be projected to the same pixel in the reference image, whatever the depth is. Therefore, each ray after pixel warping takes the ground truth color and the photometric loss will be 0.

The solution I tried is:

if self.split == 'train':
            render_idx = self.train_img_idx[idx % self.n_views]
            support_idxs = [i for i in self.train_img_idx if i != render_idx]

Even though it didn't solve the problem during fine-tuning, it works in my refactored code (based on your framework and pixel/patch-based loss).

Hope my finding could help, correct me if I'm wrong! Thanks.

―

**Comment by jerryxu9905** [comment]
> Hello. Does the "weight_sum" become 0 in fine-tuning? If so, you can adjust the two parameters "visibility_beta, and visibility_gama" used in fine-tuning.


Hello @flamehaze1115. Yes, the `weight_sum` become 0 in fine-tuing,I have tried many times for `visibility beta` and `visibility gama` params, but still can't get a reliable result, let alone every 15 scenes try shall cost so much time.... Can you share your `visibility beta` and `visibility gama` settings in every 15 test scenes? It will help us save lot of time.


Many thanks.

**Comment by flamehaze1115** [comment]
> > Hello. Does the "weight_sum" become 0 in fine-tuning? If so, you can adjust the two parameters "visibility_beta, and visibility_gama" used in fine-tuning.
> 
> Hello @flamehaze1115. Yes, the `weight_sum` become 0 in fine-tuing,I have tried many times for `visibility beta` and `visibility gama` params, but still can't get a reliable result, let alone every 15 scenes try shall cost so much time.... Can you share your `visibility beta` and `visibility gama` settings in every 15 test scenes? It will help us save lot of time.
> 
> Many thanks.

After fixing the dataloader bug, still have the problem?

**Comment by jerryxu9905** [comment]
> After fixing the dataloader bug, still have the problem?

@flamehaze1115 
Yes, I'm using the latest code where in dtu_ft.py
```python
if self.split == 'train':
            render_idx = self.train_img_idx[idx % self.n_views]
            support_idxs = [i for i in self.train_img_idx if i != render_idx]
```
shell command is:
```python
#!/usr/bin/env bash
python exp_runner_finetune.py \
--mode train --conf ./confs/finetune.conf --is_finetune \
--checkpoint_path ./weights/ckpt.pth \
--case_name scan118  --train_imgs_idx 0 1 2 --test_imgs_idx 0 1 2 --near 700 --far 1100 \
--visibility_beta 0.015 --visibility_gama 0.010 --visibility_weight_thred 0.7
```

scan118 finetune: *0 iters -> 1000iters -> 2000iters*
------
<div align="center">
<img src="https://user-images.githubusercontent.com/97662087/190587137-47910fbb-1c5b-4ab3-90c6-bb63c893ff11.png" alt="2022-09-16-15-49-59" width="50%">
<img src="https://user-images.githubusercontent.com/97662087/190587149-5af959c3-17bd-48e5-95c1-893503b29588.png" alt="2022-09-16-15-51-14" width="50%">
<img src="https://user-images.githubusercontent.com/97662087/190587152-6be51959-018d-42cf-a0a2-72357420aeac.png" alt="2022-09-16-15-51-51" width="50%">
</div>

**Comment by flamehaze1115** [comment]
Ok. I will check the code and back asap.

**Comment by zjhthu** [comment]
I find removing the `Consistency loss` term can resolve this problem, although this might give suboptimal results as shown in the paper. More specifically  `loss = torch.mean(weight[mask] * error)` in this [line](https://github.com/xxlong0/SparseNeuS/blob/main/loss/color_loss.py#L59) and `term2, term3 = 0, 0`.

**Comment by flamehaze1115** [comment]
Hello guys. Sorry for the late. I am very busy this week. I have updated the ft bash. The old parameters should be not proper. 
The consistency-aware loss is sensitive, so we introduce another parameter "visibility_weight_thred" to dynamically adjust the beta param. This param will increase the beta in the training if the "weight_sum" becomes too small.

`#!/usr/bin/env bash
python exp_runner_finetune.py \
--mode train --conf ./confs/finetune.conf --is_finetune \
--checkpoint_path ./weights/ckpt.pth \
--case_name scan118  --train_imgs_idx 0 1 2 --test_imgs_idx 0 1 2 --near 700 --far 1100 \
--visibility_beta 0.025 --visibility_gama 0.010 --visibility_weight_thred 0.7 0.6 0.5`

**Comment by FENGGENYU** [comment]
Any update on this issue? I still get empty mesh after fine-tuning, even it is from the latest bash. 

The shape mesh degenerates after fine-tuning.

![image](https://github.com/xxlong0/SparseNeuS/assets/11835443/c318e8de-971f-4297-abea-ae33cdfab52a)

**Comment by guangxuwang** [comment]
still same error
